main.go

```diff
@@ -0,0 +1,127 @@
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	metal3iov1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	metal3iocontroller "github.com/metal3-io/baremetal-operator/controllers/metal3.io"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/demo"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/fixture"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic"
+	"github.com/metal3-io/baremetal-operator/pkg/version"
+	// +kubebuilder:scaffold:imports
+)
+
+var (
+	scheme   = k8sruntime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = metal3iov1alpha1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
+
+func printVersion() {
+	setupLog.Info(fmt.Sprintf("Go Version: %s", runtime.Version()))
+	setupLog.Info(fmt.Sprintf("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH))
+	setupLog.Info(fmt.Sprintf("Component version: %s", version.String))
+}
+
+func main() {
+	var devLogging bool
+	var runInTestMode bool
+	var runInDemoMode bool
+	var metricsAddr string
+	var watchNamespace string
+
+	flag.BoolVar(&devLogging, "dev", false, "enable developer logging")
+	flag.StringVar(&metricsAddr, "metrics-addr", "127.0.0.1:8085",
+		"The address the metric endpoint binds to.")
+	// From CAPI point of view, BMO should be able to watch all namespaces
+	// in case of a deployment that is not multi-tenant. If the deployment
+	// is for multi-tenancy, then the BMO should watch only the provided
+	// namespace.
+	flag.StringVar(&watchNamespace, "namespace", "",
+		"Namespace that the controller watches to reconcile BMO objects.")
+	flag.BoolVar(&runInTestMode, "test-mode", false, "disable ironic communication")
+	flag.BoolVar(&runInDemoMode, "demo-mode", false,
+		"use the demo provisioner to set host states")
+	flag.Parse()
+
+	ctrl.SetLogger(zap.New(zap.UseDevMode(devLogging)))
+
+	printVersion()
+
+	ctrl.Log.Info(fmt.Sprintf("gather metrics at http://%s/metrics", metricsAddr))
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                  scheme,
+		MetricsBindAddress:      metricsAddr,
+		Port:                    9443,
+		LeaderElection:          true,
+		LeaderElectionID:        "baremetal-operator",
+		LeaderElectionNamespace: watchNamespace,
+		Namespace:               watchNamespace,
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	var provisionerFactory provisioner.Factory
+	if runInTestMode {
+		ctrl.Log.Info("USING TEST MODE")
+		provisionerFactory = fixture.New
+	} else if runInDemoMode {
+		ctrl.Log.Info("USING DEMO MODE")
+		provisionerFactory = demo.New
+	} else {
+		provisionerFactory = ironic.New
+		ironic.LogStartup()
+	}
+
+	if err = (&metal3iocontroller.BareMetalHostReconciler{
```

Comment:
personally not a fan of this making the Reconciler struct public, you could just make a metal3iocontroller.New()

Reply:
This is what kubebuilder generates. I don't know if the controller runtime expects these members to be public or not, but didn't want to deviate too far from what the code generator does in this PR.